src/server/pkg/collection/collection.go

```diff
@@ -438,6 +482,38 @@ func (i *iterator) Next(key *string, val proto.Unmarshaler) (ok bool, retErr err
 	return false, nil
 }
 
+// Next() writes a fully qualified key (including the path) to the key value
+func (i *paginatedIterator) Next(key *string, val proto.Unmarshaler) (ok bool, retErr error) {
+	if i.index < len(i.resp.Kvs) {
+		kv := i.resp.Kvs[i.index]
+		i.index++
+
+		*key = string(kv.Key)
+		if err := val.Unmarshal(kv.Value); err != nil {
+			return false, err
+		}
+
+		return true, nil
+	}
+	if len(i.resp.Kvs) == 0 {
+		//empty response
+		return false, nil
+	}
+	// Reached end of resp, try for another page
+	fromKey := string(i.resp.Kvs[len(i.resp.Kvs)-1].Key)
+	resp, err := i.etcdClient.Get(i.ctx, fromKey, etcd.WithRange(i.endKey), etcd.WithLimit(QueryPaginationLimit))
+	if err != nil {
+		return false, err
+	}
+	if len(resp.Kvs) == 1 {
+		// Only got the from key, there are no more kvs to fetch from etcd
+		return false, nil
+	}
+	i.index = 1 // Move past the from key
```

Comment:
This seems a bit weird to. Is it that the first key is going to always be the `from` key which we've already returned? Is this true even of the first request we send out?

Reply:
Similar deal ... because etcd returns the `fromKey` in the results (and the previous page has already provided that data through the iterator) we want to move past that key in the results we return via the iterator.